### PR TITLE
Fb rtd2

### PR DESCRIPTION
**Pull request by CarstenHansen**
This feature allows the user to specify output boundary conditions to rotated grids under switch /RTD.

Explanations are added to the manual Sect. "3.4.9 Rotated grids".

The pole of each destination grid is specified as the I'th element
of arrays BPLAT(1:9) and BPLON(1:9) in a namelist &ROTB, e.g. as
'&ROTB BPLAT(2) = 20., BPLON(2) = -130., / ! boundary conditions to nest2.ww3'

The default of the namelist &ROTB is output to standard spherical grids,
BPLAT(1:9)=90, BPLON(1:9)=-180.

### CODE REVIEW

**Comment by ukmo-ccbunney** [comment]
Hi @CarstenHansen. It looks like your branch has some merge conflicts with our develop branch. Is your branch up-to-date with the latest development from the NOAA develop branch? Looking at the commit history in this branch, I cannot see the latest commits that are present in the NOAA/UKMO develop branches...

**Comment by CarstenHansen** [comment]
Hi @ukmo-ccbunney. 

When I press the '+- Compare' link on the github webpage and choose 'base repository: ukmo-waves/WW3, base: develop', it says:
'Able to merge. These branches can be automatically merged.'

I see that I added the latest commit 8e7e264 to my branch on May 28, which was a small, trivial update. This commit has 2 parents 239b75f + 3085d9f.  3085d9f is the present NOAA-EMC develop, and 239b75f was my previous merge of NOAA-EMC develop. Would it help to revert this commit 8e7e264 to 3085d9f ?

 I wonder why I cannot view change to fb_rtd2 in commit 8e7e264 on the github web page. I think it may be because every time I do a merge of NOAA-EMC develop, it automatically creates a new commit.

**Comment by aliabdolali** [comment]
Hi @CarstenHansen 
Do we have a regression test for this feature? It's a necessity to have one to check this development and make sure it won't be affected by future developments. 
@ukmo-ccbunney I think this is a milestone, what do you think?

**Comment by CarstenHansen** [comment]
Hi @aliabdolali. I forgot to link to the issue https://github.com/NOAA-EMC/WW3/issues/68 in which I originally described this enhancement. I understood very little about regtests at that time, and @ukmo-ccbunney and @ukmo-ansaulter offered to help performing the regtests - maybe incorporate it in the regtest for RTD.

**Comment by aliabdolali** [comment]
Hi @CarstenHansen 
I made a milestone for this development on NOAA repo. 
https://github.com/NOAA-EMC/WW3/milestone/24
Please let me know if you need help with regtests preparation for RTD. I'd suggest looking at one of the regression test structure and prepare your input accordingly + info. Info contains the test description. If you have a netcdf or a large file for the input, send it to me, so I can add it to our ftp. 
We put some guides for regtest preparation on our wiki page.
https://github.com/NOAA-EMC/WW3/wiki/Checklist-for-Adding-or-Modifying-a-Regtest

**Comment by CarstenHansen** [comment]
Hi @aliabdolali 
Thank you for your advice. I think we should await the reviews by @ukmo-ccbunney and @ukmo-ansaulter. They may also come out with an example of an SMC grid with boundary conditions to a rotated  lat-lon grid. I will be on vacation for two weeks from today, and send you my input files thereafter.

File: model/ftn/w3iobcmd.ftn
Comment (ukmo-ansaulter) [inline model/ftn/w3iobcmd.ftn]: !/                  grid when nesting to rotated grid   ( version 7.xx )

This will be ready for when a milestone is available

File: model/ftn/w3updtmd.ftn
Comment (ukmo-ansaulter) [inline model/ftn/w3updtmd.ftn]: !/                  grid when nesting to rotated grid   ( version 7.xx )

This will be ready for when a milestone is available

File: model/ftn/w3updtmd.ftn
Comment (ukmo-ansaulter) [inline model/ftn/w3updtmd.ftn]: !/    13-Jun-2019 : Rotation only if POLAT<90 (C.Hansen)( version 7.xx )

This will be ready for when a milestone is available

File: model/ftn/ww3_grid.ftn
Comment (ukmo-ansaulter) [inline model/ftn/ww3_grid.ftn]: !/    24-Jun-2020 : RTD output b. c. to rotated grid.   ( version 7.xx )

File: model/nml/namelists.nml
Comment (ukmo-ansaulter) [inline model/nml/namelists.nml:8]: I think its a good idea to add in this .nml file, but in that case maybe we should start to populate this template with the namelist section included in the ww3_grid.inp template?

At the very least should probably add in the RTD namelists text?

File: manual/num/rotagrid.tex
Comment (ukmo-ansaulter) [inline manual/num/rotagrid.tex]: This section might need some re-phrasing to clarify how the boundary conditions are set up? Something along the lines of..

When nesting from a standard grid to a rotated grid model, both the outer and inner 
grids should be built with the {\code RTD} switch set when compiling executables for 
the models. When nesting from a rotated grid to a standard grid model, the inner 
(standard) grid model does not require to be built with {\code RTD}.

Output of spectra at boundary points to one-way nested inner grids are transferred
as described in Appendix~\ref{app:nest}. Output b.c.\ are defined in the input
file (see {\bf ww3\_grid.inp}, Sect.~\ref{sec:config011}) as a sequence of straight
lines given in coordinates of the inner grid. If the inner grid is rotated, it's
pole position must be defined as the values of the array elements
{\code BPLAT(\sl{n})} and {\code BPLON(\sl{n})} under namelist {\code ROTB}
for the outer grid. The array index {\sl{n}} is the index ({\sl{1:9}}) of the boundary 
conditions file {\file nest{\sl{n}}.ww3}.

**Review by ukmo-ansaulter — COMMENTED** [review]
Thanks for doing such a great job @CarstenHansen 

I've tested these changes with our standard grid global model -> rotated grid regional model set-up and everything works OK. 

In addition to the comments I've added here one further suggested change would be to up the precision of the error outputs associated with BCs; one issue I did have was in diagnosing where I'd set the BCs up incorrectly onto a high resolution grid. The suggested change is for lines 6051-52

  994 FORMAT ( '   *** POINT OUTSIDE GRID (SKIPPED) : X,Y =',2F10.5)
  995 FORMAT ( '   *** POINT ON LAND      (SKIPPED) : X,Y =',2F10.5)

Regarding the regtest, @ukmo-ccbunney need to have a look at what to do and whether the existing tp2.11 is fit for purpose. So, if you are happy with the proposed changes, suggest you mgiht deal with these now whilst we come up with a plan for the regtest.

**Comment by CarstenHansen** [comment]
Great to see that your tests were OK @ukmo-ansaulter. 
I will adopt your proposed changes. Regarding the need for a high resolution of the error message for points outside the grid or on land, I think that if anybody uses Cartesian coordinates (Easting/Northing), they should see the error output in units of m. The format descriptors 2994, 2995 are there and are brought into use with the construction:

                IF ( FLAGLL ) THEN
                    WRITE (NDSE,2994) FACTOR*XO, FACTOR*YO
                  ELSE
                    WRITE (NDSE,994) FACTOR*XO, FACTOR*YO
                  END IF

File: model/ftn/w3iobcmd.ftn
Comment (CarstenHansen) [reply model/ftn/w3iobcmd.ftn]: ok

File: model/ftn/w3updtmd.ftn
Comment (CarstenHansen) [reply model/ftn/w3updtmd.ftn]: ok

File: model/ftn/w3updtmd.ftn
Comment (CarstenHansen) [reply model/ftn/w3updtmd.ftn]: ok

File: model/ftn/ww3_grid.ftn
Comment (CarstenHansen) [reply model/ftn/ww3_grid.ftn]: Ok, and also changing resolution of error output as suggested.

File: model/nml/namelists.nml
Comment (CarstenHansen) [reply model/nml/namelists.nml:8]: I have included the namelist section of the ww3_grid.inp template and the RTD namelists text. I edited the ROTB text in ww3_grid.inp to make it consistent.

File: manual/num/rotagrid.tex
Comment (CarstenHansen) [reply manual/num/rotagrid.tex]: This is a good clarification. I consider the manual  the most valuable component of the WW3 distribution.

**Review by ukmo-ansaulter — COMMENTED** [review]
Thanks for adopting these changes @CarstenHansen 

In terms of the code, I approve this review. Will finalise an approval once @ukmo-ccbunney has revised and tested the new RTD regtests

**Comment by ukmo-ccbunney** [comment]
@CarstenHansen I have some regression tests ready for this change; we need to get them added to your branch. You can either:
   a) Give me write access to your branch on GitHub and I will push the changes up, or:
   b) I can put the changes on a separate branch in this repo and you can merge them in.
Let me know what you prefer.

**Comment by CarstenHansen** [comment]
> a) Give me write access to your branch on GitHub and I will push the changes up, or:

@ukmo-ccbunney I have invited you as a collaborator, which is the way i seems you will have write (push) access.

**Comment by ukmo-ccbunney** [comment]
>I have invited you as a collaborator

Thanks @CarstenHansen - I've pushed a commit to your branch containing the regtests (see ww3_tr1).

There are 3 tests in seperate input directories (see the `info` file for more details). Basically, the first two tests (`input`, `input_std`) test the generation of output boundary conditions on target grids formulated on different poles. The difference between the tests is that for the first the source grid is formulated on a rotated pole, and the second is on a standard pole.

The third test under the `input_bndin` technically does not test your RTD2 changes - it tests the correct ingestion of input `nest.ww3` files on to a rotated pole model. I have included it in this regtest set as it complements the other two tests. Technically, it requires the update to the `ww3_bound` program that is handled in a separate PR #10, although it should still run without errors (it just gets the distance calculations to the points wrong).

**Comment by CarstenHansen** [comment]
Thank you. Your explanation (and the file regtests/ww3_tr1/info) makes good sense to me.

**Review by ukmo-ansaulter — APPROVED** [review]
Approving this request - the new regtests are appropriate for covering RTD functionality

**Comment by ukmo-ccbunney** [comment]
@CarstenHansen : I've just pushed up commit https://github.com/ukmo-waves/WW3/pull/7/commits/9e24969bef587bb53595cb8a8a9591e4b38d681d which contains a merge of the develop branch. There were a few changes required to fix some merge conflicts. I have also updated the version numbers inline with the milestone for this change.

Can you just check you are happy with the modifications?

**Comment by CarstenHansen** [comment]
Hi @ukmo-ccbunney. Commit 9e24969 looks fine, I just found one typo in model/ftn/w3_bound.ftn: 'version 7.06' should be 'version 7.11' ?

**Comment by ukmo-ccbunney** [comment]
> 
> 
> Hi @ukmo-ccbunney. Commit [9e24969](https://github.com/ukmo-waves/WW3/commit/9e24969bef587bb53595cb8a8a9591e4b38d681d) looks fine, I just found one typo in model/ftn/w3_bound.ftn: 'version 7.06' should be 'version 7.11' ?

Hi @CarstenHansen - I originally did not consider the ww3_bound/ww3_bounc changes to be part of the 7.11 milestone as they were more bugfixes to the original RTD functionality. However, it seems that the ww3_bounc.ftn module has been updated to 7.11, so let's keep the ww3_bound.ftn module the same and make it 7.11 too. I'll push up a modification. Thanks.

**Comment by ukmo-ccbunney** [comment]
Regtest for `ww3_tp2.11` (rotated pole regression test) shows no differences, except for the expected extra screen output from `ww3_grid`:

```
***
/home/d02/frey/WW3/CarstenRTD/regtests/output/ww3_tp2.11/work_SHRD_RTD/ww3_grid.out_diff.txt
***
146,148d145
<   &ROTD PLAT = 37.50, PLON = 177.50, UNROT =  F /
<   &ROTB BPLAT =  90.0,  90.0,  90.0,  90.0,  90.0,  90.0,  90.0,  90.0,  90.0,
<         BPLON =-180.0,-180.0,-180.0,-180.0,-180.0,-180.0,-180.0,-180.0,-180.0, /
```

I am going to run the full regression test suite now.

**Comment by ukmo-ccbunney** [comment]
Results of full regression test matrix: [matrixComp.zip](https://github.com/ukmo-waves/WW3/files/5161124/matrixComp.zip)

Summary of differences:

  - `ww3_tp2.11` : **Expected** differences in screen output of ww3_grid related to this PR (now shows new ROTB namelist defaults)
  - `mww3_test_03`: **Existing issue:** known to be not b4b.
  - `ww3_tp2.18/work_TIDE_MPI` : **Existing issue:** known differences in screen output of ww3_grid.
  - `ww3_ta1` and `ww3_ts1` : **Existing issue:** known differences in the ST4 tests - see issue https://github.com/NOAA-EMC/WW3/issues/249

I'm happy for this change to be merged.

@aliabdolali - are you happy with the changes?

**Review by ukmo-ccbunney — APPROVED** [review]
All looks good to me now.

**Comment by ukmo-ccbunney** [comment]
Just one final small commit - I forgot to add the `ww3_tr1` test to the the `all` section of `matrix.comp`.
This does not affect the regtest results above as it is a new regtest directory created in this PR.

**Review by aliabdolali — COMMENTED** [review]
It looks ok to me. Thanks for doing this clean work.

**Comment by aliabdolali** [comment]
I wanted to press approve bottom but pressed the comment bottom instead :) Please consider it approved.

**Pull request merged**
